### PR TITLE
 [ENH] Add whole-series ROCKAD anomaly detector

### PR DESCRIPTION
**Pull request by pattplatt**
#### Reference Issues/PRs

Extends on #2326 and replaces #2795 (rebased and updated to current main)

#### What does this implement/fix? Explain your changes.

Implements whole-series ROCKAD as semi-supervised anomaly detector

##### For all contributions
- [X] I've added myself to the [list of contributors](https://github.com/aeon-toolkit/aeon/blob/main/.all-contributorsrc). Alternatively, you can use the [@all-contributors](https://allcontributors.org/docs/en/bot/usage) bot to do this for you **after** the PR has been merged.
- [X] The PR title starts with either [ENH], [MNT], [DOC], [BUG], [REF], [DEP] or [GOV] indicating whether the PR topic is related to enhancement, maintenance, documentation, bugs, refactoring, deprecation or governance.

##### For new estimators and functions
- [X] I've added the estimator/function to the online [API documentation](https://www.aeon-toolkit.org/en/latest/api_reference.html).
- [ ] (OPTIONAL) I've added myself as a `__maintainer__` at the top of relevant files and want to be contacted regarding its maintenance. Unmaintained files may be removed. This is for the full file, and you should not add yourself if you are just making minor changes or do not want to help maintain its contents.

##### For developers with write access
- [ ] (OPTIONAL) I've updated aeon's [CODEOWNERS](https://github.com/aeon-toolkit/aeon/blob/main/CODEOWNERS) to receive notifications about future changes to these files.

### CODE REVIEW

**Comment by pattplatt** [comment]
As I understand it some tests are failing because the series as well as the collection ROCKAD are named the same? Should we give unique names like ROCKAD_Series and ROCKAD_Collection? However this would not fit in the current naming schema... Or can the tests be updated?

**Comment by MatthewMiddlehurst** [comment]
Hi, I fixed this in a separate PR. It may be better to rename one of them either way, i.e. SubsequenceROCKAD or something

**Comment by pattplatt** [comment]
> Hi, I fixed this in a separate PR. It may be better to rename one of them either way, i.e. SubsequenceROCKAD or something

Ok, should we then just rename the whole-series approach to class WholeSeriesRockad()? File name can stay the same (_rockad)?

**Comment by MatthewMiddlehurst** [comment]
I will leave it up to you, I have no preference. I assumed the whole series one was the original.

**Comment by pattplatt** [comment]
> I will leave it up to you, I have no preference. I assumed the whole series one was the original.

I’ve thought about this again, and if the test issue has been resolved anyway, I would actually prefer keeping the original class names. This way, the class names across the different ML approaches remain consistent. In my opinion, the abstraction between BaseSeriesAnomalyDetector and BaseCollectionAnomalyDetector is already clear enough.

**Comment by SebastianSchmidl** [comment]
I am also fine with keeping the names the same. @MatthewMiddlehurst do you prefer separate names, anyway?

**Comment by TonyBagnall** [comment]
sorry, missed the PR, I dont mind if they have the same name either

File: aeon/anomaly_detection/collection/_rockad.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+
+        super().__init__()
+
+    def _fit(self, X: np.ndarray, y: Optional[np.ndarray] = None) -> "ROCKAD":
+        _X = X
+        self._inner_fit(_X)
+
+        return self
+
+    def _inner_fit(self, X: np.ndarray) -> None:
```
Comment (TonyBagnall) [inline aeon/anomaly_detection/collection/_rockad.py]: why do you need the two functions? Just refactor _inner_fit as _fit?

File: aeon/anomaly_detection/collection/_rockad.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+        "capability:multivariate": True,
+        "capability:missing_values": False,
+        "capability:multithreading": True,
+        "fit_is_empty": False,
+    }
+
+    def __init__(
+        self,
+        n_estimators=10,
+        n_kernels=100,
```
Comment (TonyBagnall) [inline aeon/anomaly_detection/collection/_rockad.py:97]: given the internal type hints, may as well have type hints on the constructor argument?

File: aeon/anomaly_detection/collection/_rockad.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+        # XT: (n_cases, n_kernels*2)
+        Xt = self.rocket_transformer_.fit_transform(X)
+        Xt = Xt.astype(np.float64)
+
+        if self.power_transform:
+            self.power_transformer_ = PowerTransformer()
+            try:
+                Xtp = self.power_transformer_.fit_transform(Xt)
+
+            except Exception:
```
Comment (TonyBagnall) [inline aeon/anomaly_detection/collection/_rockad.py]: why does power_transform fail? Perhaps also print out the thrown exception in the warning?

File: aeon/anomaly_detection/collection/_rockad.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+                n_samples=None,
+                random_state=self.random_state + idx_estimator,
+                stratify=None,
+            )
+
+            # Fit estimator and append to estimator list
+            estimator.fit(Xtp_scaled_sample)
+            self.list_baggers_.append(estimator)
+
+    def _predict(self, X) -> np.ndarray:
```
Comment (TonyBagnall) [inline aeon/anomaly_detection/collection/_rockad.py:164]: same as above, dont really need predict and inner predict?

**Review by TonyBagnall — DISMISSED** [review]
couple of structural points, fine otherwise

**Comment by MatthewMiddlehurst** [comment]
Hi, i reworked how to thread a bit so there was a new failure here. I have fixed this a few small bits myself.

File: aeon/anomaly_detection/collection/_rockad.py [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+    def __init__(
+        self,
+        n_estimators=10,
+        n_kernels=100,
+        normalise=False,
+        n_neighbors=5,
+        metric="euclidean",
+        power_transform=True,
+        n_jobs=1,
+        random_state=None,
```
Comment (MatthewMiddlehurst) [inline aeon/anomaly_detection/collection/_rockad.py:103]: Changed this to no seed, which is our usual default

File: aeon/anomaly_detection/collection/_rockad.py [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+        self.random_state = random_state
+
+        self.rocket_transformer_: Optional[Rocket] = None
+        self.list_baggers_: Optional[list[NearestNeighbors]] = None
+        self.power_transformer_: Optional[PowerTransformer] = None
+
+        super().__init__()
+
+    def _fit(self, X: np.ndarray, y: Optional[np.ndarray] = None):
+        self._n_jobs = check_n_jobs(self.n_jobs)
```
Comment (MatthewMiddlehurst) [inline aeon/anomaly_detection/collection/_rockad.py:122]: Changed this to use `check_n_jobs`.

I agree with Tony re: the inner methods so just merged those with fit/predict

File: aeon/anomaly_detection/collection/tests/test_rockad.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+def test_rockad_incorrect_input():
+    """Test ROCKAD with invalid inputs."""
+    rng = check_random_state(seed=2)
+    series = rng.normal(size=(10, 5))
+
+    with pytest.warns(
+        UserWarning, match=r"Power Transform failed and thus has been disabled."
+    ):
+        ad = ROCKAD()
+        ad.fit(series)
```
Comment (MatthewMiddlehurst) [inline aeon/anomaly_detection/collection/tests/test_rockad.py]: This seems to be failing (or inconsistently passing)?

File: aeon/anomaly_detection/collection/_rockad.py [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+                metric=self.metric,
+                algorithm="kd_tree",
+            )
+            # Bootstrap Aggregation
+            Xtp_scaled_sample = resample(
+                Xtp,
+                replace=True,
+                n_samples=None,
+                random_state=rng.randint(np.iinfo(np.int32).max),
+                stratify=None,
```
Comment (MatthewMiddlehurst) [inline aeon/anomaly_detection/collection/_rockad.py:157]: Changed to use a `RandomState` for rng

File: aeon/anomaly_detection/collection/_rockad.py [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+    """
+    ROCKET-based whole-series Anomaly Detector (ROCKAD).
+
+    ROCKAD [1]_ leverages the ROCKET transformation for feature extraction from
+    time series data and applies the scikit learn k-nearest neighbors (k-NN)
+    approach with bootstrap aggregation for robust semi-supervised anomaly detection.
+    The data gets transformed into the ROCKET feature space.
+    Then the whole-series are compared based on the feature space by
+    finding the nearest neighbours. The time-point based ROCKAD anomaly detector
+    can be found at aeon/anomaly_detection/series/distance_based/_rockad.py
```
Comment (SebastianSchmidl) [inline aeon/anomaly_detection/collection/_rockad.py]: This does not render as a proper code link in the docs

File: aeon/anomaly_detection/collection/_rockad.py (hunk 3 above)
Comment (SebastianSchmidl) [reply aeon/anomaly_detection/collection/_rockad.py]: At least we should add a comment in the code explaining why this is necessary.

**Comment by MatthewMiddlehurst** [comment]
I'm not sure the PowerTransform bit is necessary and I can't seem to reproduce it with a test so have removed it for now. Maybe @pattplatt can give some help here if it is necessary. It can be turned off with a parameter in either case for the same functionality as the automatic disabling.

Just removed the docs bit for now, they both have the same name so not sure if that will cause issues when linking to each other in a "See Also" section or whatever.

File: aeon/anomaly_detection/collection/_rockad.py (hunk 2 above)
Comment (SebastianSchmidl) [reply aeon/anomaly_detection/collection/_rockad.py:97]: The types can be inferred from the default value, so they are not strictly necessary. However, for documentational purposes, I would agree.
